Objects/tupleobject.c

```diff
@@ -1020,13 +1020,15 @@ tupleiter_next(PyObject *self)
         return NULL;
     assert(PyTuple_Check(seq));
 
-    if (it->it_index < PyTuple_GET_SIZE(seq)) {
-        item = PyTuple_GET_ITEM(seq, it->it_index);
-        ++it->it_index;
+    Py_ssize_t idx = FT_ATOMIC_LOAD_SSIZE_RELAXED(it->it_index);
+    if ((size_t)idx < (size_t)PyTuple_GET_SIZE(seq)) {
+        item = PyTuple_GET_ITEM(seq, idx++);
+        FT_ATOMIC_STORE_SSIZE_RELAXED(it->it_index, idx);
         return Py_NewRef(item);
     }
 
-    it->it_seq = NULL;
+    FT_ATOMIC_STORE_SSIZE_RELAXED(it->it_index, -1);
+    FT_ATOMIC_STORE_PTR_RELAXED(it->it_seq, NULL);
```

Comment:
We should not set `it_seq` to `NULL` in the free threaded build. It's not safe with concurrent calls to `tupleiter_next`.

Reply:
Well, I have PR #128637 to fix that, and I want to get that in first. (That's why this is still a draft PR really.)